Bluetooth: tester: Fix undeclared CONFIG_UART_PIPE_ON_DEV_NAME

MariuszSkamra opened the pull request:
This fixes build error due to undeclared CONFIG_UART_PIPE_ON_DEV_NAME
on nrf52840 board.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

jhedberg commented:
I was getting this error on all nRF52 boards that I tried. It looks like this will only fix it for one of them?

jhedberg commented:
Also, it seems odd that this would have to be fixed for every application wanting to use CONFIG_UART_PIPE_ON_DEV_NAME, considering that it's not a single-app feature (if it was then Kconfig is the wrong place for it). So isn't there some better non-app-specific place to fix this?

jhedberg reviewed (changes requested):
See my two earlier comments

MariuszSkamra commented:
jhe: I could set this in boards/arm/nrf*/*.dts, but this could conflict with zephyr,console = &uart0, see:
`/ {
	model = "Nordic PCA10056 Dev Kit";
	compatible = "nordic,pca10056-dk", "nordic,nrf52840-qiaa",
		     "nordic,nrf52840";

	chosen {
		zephyr,console = &uart0;
		zephyr,sram = &sram0;
                zephyr,uart-pipe = &uart0;
	};
};`

jhedberg commented:
@MariuszSkamra I have no idea. Perhaps @SebastianBoe has a suggestion for how to fix this cleanly?

SebastianBoe commented:
> @jhedberg I could set this in boards/arm/nrf*/*.dts, but this could conflict with zephyr,console = &uart0, see:

It doesn't conflict with console because it is not valid to enable console and uart-pipe simultaneously.

> Is it ok?

Yes.

PS: I don't really get why DTS is used to select which peripheral a SW component uses at all. I don't have a good understanding of Kconfig and dts yet, but you would think that a user should be able to kconfigure which peripheral a SW component is using.

MariuszSkamra commented:
There are many boards that will fail to build uart-pipe. I'm not sure I should provide a patch that affects every board... I will push a change only for nrf52840, because this is the one I can test patch with.

MariuszSkamra commented:
updated

SebastianBoe commented:
What confuses me is that the kconfig already looks like this:

```
if ! HAS_DTS
config UART_PIPE_ON_DEV_NAME
	string "Device Name of UART Device for pipe UART"
	default "UART_0"
	depends on UART_PIPE
	help
	This option specifies the name of UART device to be used
	for pipe UART.
endif

```

so you would think that was already defaulting to UART_0 without the proposed patch. I am
struggling to see why UART_PIPE_ON_DEV_NAME is not defined as UART_0 before this patch.

galak commented:
I need to look at the details, but I think a dts overlay might help with this.

MariuszSkamra commented:
> @SebastianBoe What confuses me is that the kconfig already looks like this:
> if ! HAS_DTS
> so you would think that it was already defaulting to UART_0 without the proposed patch. I am
> struggling to see why UART_PIPE_ON_DEV_NAME is not defined as UART_0 before this patch.

Note that there is negation operator. This board (target) supports device tree config, so HAS_DTS is set to true.

@galak Yes, overlay does the work, but it can be applied only per board. Am I right?

SebastianBoe commented:
> Note that there is negation operator. 

My mistake ...

galak commented:
> Note that there is negation operator. This board (target) supports device tree config, so HAS_DTS is set to true. UART_PIPE_ON_DEV_NAME won't be set.
> 
> @galak Yes, overlay does the work, but it can be applied only per board. Am I right?

We could improve things to introduce a non-board specific overlay, but I question if that truly makes sense.  Do you know that the UART is always going to be the same regardless of the board?

cvinayak commented:
I agree with @galak that samples that use UART_PIPE have overlay files (reason being, we dont want users enabling console and uart-pipe at the same time when there is only one supported uart peripheral).

MariuszSkamra committed:
Bluetooth: tester: Fix uart-pipe build error on nrf52840 board

This fixes build error due to undeclared CONFIG_UART_PIPE_ON_DEV_NAME
on nrf52840 board.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
OK, as agreed, I pushed overlay file for this board for tester application.
@galak, @cvinayak thanks for suggestion
